EEPROM + LoRa implementation

RoniNuutinen opened the pull request:
EEPROM and LoRa functionality implemented and working.

phuong-ta committed:
Lora init and time done

phuong-ta committed:
lora modified

phuong-ta committed:
BTN created!

phuong-ta committed:
LED added!

phuong-ta committed:
connect_network and send_msg functions for lora done!

RoniNuutinen committed:
rearranged some files to operation

RoniNuutinen committed:
Merge branch 'dev-master' of https://github.com/hannahhoang2704/Pill_dispenser into roni

RoniNuutinen committed:
Merging dev-master stuff with my operation setup

RoniNuutinen committed:
pulled Phuongs LoRa stuff; made some adjustments for implementation

RoniNuutinen committed:
version number 7632 of LED setup

RoniNuutinen committed:
the flags returned to their birth place

RoniNuutinen committed:
ze flag

RoniNuutinen committed:
version number 187253; changed switches to the same 'format' as LED's, and also for possible second switch to provide Keijo his logs

RoniNuutinen committed:
deleted obsolete macros

RoniNuutinen committed:
lots of stuff; preparing for adv. req. implementation; lots of it still under construction

RoniNuutinen committed:
forgot some minor aesthetics

RoniNuutinen committed:
working towards adv. req.. This is a working version, just not yet 'advanced'

RoniNuutinen committed:
added LoRa files into executables

RoniNuutinen committed:
eeprom files merged from dev-master

RoniNuutinen committed:
minor changes + lora rsp's do not apparently include newlines contrary to what the documentation says

RoniNuutinen committed:
minor changes

RoniNuutinen committed:
forgot to change the functions to work with the new struct format

RoniNuutinen committed:
continued lora + eeprom addition: LoRa and EEPROM both now work ; a very crued version of logging in various places in program -- still under construction

RoniNuutinen committed:
LoRa + EEPROM implementation; a crude version ; still under construction

RoniNuutinen committed:
Minor changes: Both LoRa and EEPROM had STRLEN macro, with different values; now distinquished. Added control over function behaviour. Added a fancy loading bar.

RoniNuutinen committed:
Both EEPROM and LoRa had STRLEN macro, with different values; now distinquished.

RoniNuutinen committed:
logf_msg is here! Does it all, just give it a million parameters and it will solve world hunger.

RoniNuutinen committed:
Added read_log at the end of the main loop.

RoniNuutinen committed:
Added 'failure response messages' for faster confirmation, and minor printf feedback.

RoniNuutinen committed:
Created a new function for simpler value calling from registries -- used for single byte reading.

RoniNuutinen committed:
Upgraded logf_msg with a time stamp, called from get_time_with_decimal. Redid operation initialization.

RoniNuutinen committed:
Changed order of initializations. Nothing huge.

hannahhoang2704 commented on operation/operation.c:
```diff
@@ -0,0 +1,245 @@
+    va_start(args, n_args);
+    vsnprintf(content, STRLEN_EEPROM - 10, log_msg[logEnum], args);
+    va_end(args);
+    strcat(msg, content);
```
I would prefer to use strncat instead of strcat. even though I understand that you've use snprintf to be sure that only 9 characters from time will be passed in log:)

hannahhoang2704 committed:
define some state names

hannahhoang2704 commented on eeprom/eeprom.c:
```diff
@@ -118,17 +123,15 @@ void read_log_entry(uint8_t *index){
-        write_to_eeprom(write_address, buf, ENTRY_SIZE);
+        write_to_eeprom(write_address, buf, 1);
     }
+    *log_index = FIRST_ADDRESS;
```
log index is not the same as FIRST ADDRESS even though the value is both 0.

RoniNuutinen committed:
Simple comment explanation for board_index calculation

RoniNuutinen committed:
A clean up attempt. Hopefully easier to read / implement / modify.

RoniNuutinen committed:
strNcat to ensure no overflows

RoniNuutinen committed:
Reverted index reset back to 0

RoniNuutinen committed:
Merged my own branch into my changes

hannahhoang2704 commented on operation/operation.h:
```diff
@@ -6,57 +6,70 @@
+    COMPARTMENT_ROTATION_STARTED,
+    COMPARTMENT_ROTATION_FINISHED,
+    PILL_DETECTED,
+    PILL_NOT_DETECTED,
```
I've uploaded another commit that clarify this part. We're now redo the work. plz check my version before make any further adjustment.

hannahhoang2704 commented:
please check my commit that I have fixed strncat, rename states and reset index. [e468545]
We're now doing duplicated work. The comment I left before is to provide reason of the code change in the commit that I pushed. So please  revert your commit if possible and give comment on my commit. thanks

RoniNuutinen commented:
> please check my commit that I have fixed strncat, rename states and reset index. [[e468545](https://github.com/hannahhoang2704/Pill_dispenser/commit/e468545496bbeeb5ea3f7b34dc8d1550795b7370)] We're now doing duplicated work. The comment I left before is to provide reason of the code change in the commit that I pushed. So please revert your commit if possible and give comment on my commit. thanks

Oh...
I also changed the logf_msg in hopes to make it easier to read,
as well as modified the switch operations to adhere to logging the same way everything else does.

I could also cancel this pull request,
wait for your irq changes,
pull them from dev-master to roni and resolve the conflicts,
then pull request my changes again.

hannahhoang2704 commented:
the last commit is already merged between your commit and my commit so plz keep this pull request, don't cancel it. the irq combination should be handled separately. Let's not extend this pull request any longer and plz avoid duplicated work with different commit next time.

RoniNuutinen commented:
> Let's not extend this pull request any longer and plz avoid duplicated work with different commit next time.

I mean I don't know why this is still hanging, to be honest...
I can agree that it isn't perfect, but that's what dev-master is for; this is still work in progress.

> The irq combination should be handled separately.

Do you mean that I shouldn't pull the irq-stuff into roni even after having merged them to dev-master?

At the moment the roni and irq-combine have a conflict ; one of them needs to resolve the other for them to coexist in dev-master, no?
Another alternative:
merge one of the two to dev-master,
then manually pull the other to dev-master and resolve their conflict there.

hannahhoang2704 commented:
the irq combine use roni branch as base to start as a separate branch, later if it's accepted, it will be merged to dev-master, not roni. I'll handle rebase the irq combine branch and merge to dev-master later. same with this roni branch it should be merged to dev-master. Dev-master and main are the official branches that we keep our 'final' code.

RoniNuutinen committed:
Incorrect switch case fixed.
